tasks/hana_deploy.yml

```diff
@@ -40,21 +40,106 @@
   register: sap_hana_deployment_register_hdblcm_stat
   failed_when: not sap_hana_deployment_register_hdblcm_stat.stat.exists
 
-- name: create temporary directory to store the processed template
-  tempfile:
-    state: directory
-    suffix: hanaconfig
-  register: tmpdir
-
-- name: "Process HANA Configfile Template"
-  template:
-    src: "{{ role_path }}/templates/configfile.j2"
-    dest: "{{ tmpdir.path }}/configfile.cfg"
-  register: cftemplate
-
-- name: Install SAP HANA
-  command: "./hdblcm {{ sap_hana_deployment_hdblcm_extraargs }} --configfile={{ tmpdir.path }}/configfile.cfg -b"
-  register: installhana
-  args:
-    chdir: "{{ sap_hana_installdir }}"
-  changed_when: "'SAP HANA Lifecycle Management' in installhana.stdout"
+- name: Search for installed SAP Systems
+  shell: |
+         if [ -x /usr/sap/hostctrl/exe/saphostctrl ]; then
```

Comment:
This is not the way to do this in Ansible. Pleae consider moving this to 'stat' module or similar.

Reply:
I thought about it and would make the playbook less readable and more complex as I have to replace this with 4 different tasks to gain the same result. I think this is better readable and reduces complexity. Keep it simple.

Reply:
> Please check my comments.
> 
> I stopped the review already when I saw the changes in hana_deploy.yml. Generally speaking, I don't see the value of adding such complexity to the role to check the existing installations and show this information to the user, I suggest you move this to a different '2-day operation'-like role.
> 
I totally disagree, a role should ALWAYS be idempotent. the current role is NOT idempotent.  If you rerun this role on a system, a second time the role fails because HANA is already there and running. With my fixes the role reports back, that HANA is already running with desired SID & Instance number, installs it if it is not there, or it fails before it actually calls the installer with a proper message. The different outputs I do are helpful in debugging. Given the case the role fails, you can immediately see what's on the system.

> As I mentioned in the related issue #10 the role will already fail without hurting existing installation if the SID already exists in the system. If you want to specifically control this behaviour I'd suggest to just add this particular check and stop the role execution with a custom message.

That's what I do. But I stop the role without an error, if the correct stuff is running

Reply:
> I thought about it and would make the playbook less readable and more complex as I have to replace this with 4 different tasks to gain the same result. I think this is better readable and reduces complexity. Keep it simple.

I do not agree. Using specific modules will give the playbook idempotence and provide better portability and readability. What you are doing here is not an Ansible best practice, check point 4 from [this](https://www.ansible.com/blog/ansible-best-practices-essentials) article.

Reply:
> > Please check my comments.
> > I stopped the review already when I saw the changes in hana_deploy.yml. Generally speaking, I don't see the value of adding such complexity to the role to check the existing installations and show this information to the user, I suggest you move this to a different '2-day operation'-like role.
> 
> I totally disagree, a role should ALWAYS be idempotent. the current role is NOT idempotent. If you rerun this role on a system, a second time the role fails because HANA is already there and running. With my fixes the role reports back, that HANA is already running with desired SID & Instance number, installs it if it is not there, or it fails before it actually calls the installer with a proper message. The different outputs I do are helpful in debugging. Given the case the role fails, you can immediately see what's on the system.
> 
> > As I mentioned in the related issue #10 the role will already fail without hurting existing installation if the SID already exists in the system. If you want to specifically control this behaviour I'd suggest to just add this particular check and stop the role execution with a custom message.
> 
> That's what I do. But I stop the role without an error if the correct stuff is running

I think you are misunderstanding the idempotency concept here. This role will NEVER be idempotent since we are relying on `hdblcm` tool from SAP to ensure HANA is installed. Since `hdblcm` is not idempotent, the role won't either. Modifying the output message of a failure in the role doesn't make a role idempotent.

Also please check the actual output from `hdblcm` I added to the issue related to this PR:

```
"stdout": "\n\nSAP HANA Lifecycle Management - SAP HANA Database 2.00.045.00.1575639312\n************************************************************************\n\n\nScanning software locations...\nDetected components:\n    SAP HANA Database (2.00.045.00.1575639312) in /usr/local/src/SAP_HANA_DATABASE/server\nLog file written to '/var/tmp/hdb_SLE_hdblcm_install_2020-01-22_05.44.20/hdblcm.log' on host 'rhel-target-02'.",
    "stderr": "Configuration error (batch mode):\n  SAP HANA System ID 'SLE' is already in use by a SAP HANA Database instance!",
    "rc": 1,
```

For me, this is clear enough for the user, if you want to make it this more "readable", my suggestion is the following, so you don't overcomplicate the actual role as you are doing:

- Create a new playbook called `tasks/pre_checks.yml` for example.
- Modify `tasks/main.yml` to import tasks from the newly created playbook as the first thing to do.
- Add the logic of checking for the SID and fail the role with a custom message to this newly created playbook.
- Leave the rest of the role as-is

Reply:
I made these changes now to tasks. I have not yet tested, but will do soon